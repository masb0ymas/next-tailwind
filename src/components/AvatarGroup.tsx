/* eslint-disable @next/next/no-img-element */
import { AvatarEntity } from '@src/data/entities/picture'

export type AvatarSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl'

interface AvatarGroupProps {
  data: AvatarEntity[]
  limit?: number
  size?: AvatarSize | (string & {})
}

function AvatarGroup(props: AvatarGroupProps) {
  const { data, size, limit = 4 } = props

  const newData = data.slice(0, limit)
  const remainingData = data.slice(limit)

  let newSize = ''
  let newFontSize = ''

  switch (size) {
    case 'xs':
      newSize = 'h-10 w-10'
      newFontSize = 'text-md'
      break

    case 'sm':
      newSize = 'h-12 w-12'
      newFontSize = 'text-lg'
      break

    case 'md':
      newSize = 'h-14 w-14'
      newFontSize = 'text-xl'
      break

    case 'lg':
      newSize = 'h-16 w-16'
      newFontSize = 'text-2xl'
      break

    case 'xl':
      newSize = 'h-20 w-20'
      newFontSize = 'text-3xl'

      break

    default: // default case 'md'
      newSize = 'h-14 w-14'
      newFontSize = 'text-xl'
      break
  }

  if (!newData || !newData.length) {
    return <div>No Avatars</div>
  }

  return (
    <div className="flex justify-center -space-x-4 overflow-hidden">
      {newData &&
        newData.map((row) => {
          const uiAvatar = 'https://ui-avatars.com/api/?name='
          let newSource = row.source

          if (!row.source) {
            newSource = `${uiAvatar}${row.label}`
          }

          return (
            <img
              key={row.id}
              className={`inline-block ${newSize} rounded-full ring-2 ring-white`}
              src={newSource}
              alt={row.label}
              data-testid="avatar-image"
            />
          )
        })}

      {/* check remaining avatar */}
      {remainingData.length > 0 && (
        <div
          className={`relative inline-flex ${newSize} items-center justify-center overflow-hidden rounded-full bg-gray-200 ring-2 ring-white dark:bg-gray-200`}
        >
          <span
            className={`${newFontSize} font-medium text-gray-600 dark:text-gray-600`}
            data-testid="remaining-avatar"
          >
            {`+${remainingData.length}`}
          </span>
        </div>
      )}
    </div>
  )
}

export default AvatarGroup
