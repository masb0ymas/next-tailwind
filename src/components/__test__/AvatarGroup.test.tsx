import dummyData from '@src/data/dummy/picture.json'
import '@testing-library/jest-dom'
import { cleanup, render } from '@testing-library/react'
import AvatarGroup from '../AvatarGroup'

afterEach(cleanup)

describe('Test Avatar Group', () => {
  test('renders "no avatars" when there are no avatars', () => {
    const { getByText } = render(<AvatarGroup data={[]} />)

    expect(getByText(/no avatars/i)).toBeInTheDocument()
  })

  test('renders avatars', () => {
    const newData = dummyData.data

    const { getAllByTestId } = render(
      <AvatarGroup data={newData} limit={newData.length} />
    )

    const avatarAttr = getAllByTestId('avatar-image').map((x) =>
      x.getAttribute('alt')
    )

    const avatarLists = avatarAttr.map((name) => name)
    const avatarData = newData.map((row) => row.label)

    expect(avatarLists).toEqual(avatarData)
  })

  test('remaining avatar', () => {
    const newData = dummyData.data
    const limit = 4

    const remainingData = newData.slice(limit)
    const expectResult = `+${remainingData.length}`

    const { getByTestId } = render(<AvatarGroup data={newData} limit={limit} />)
    const avatarRemaining = getByTestId('remaining-avatar').textContent

    expect(expectResult).toEqual(avatarRemaining)
  })
})
