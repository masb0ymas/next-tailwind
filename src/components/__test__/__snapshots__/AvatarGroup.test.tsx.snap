// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Avatar Group snapshot component 1`] = `
<div
  className="flex justify-center -space-x-4 overflow-hidden"
>
  <img
    alt="Jack Sparo"
    className="inline-block h-14 w-14 rounded-full ring-2 ring-white"
    data-testid="avatar-image"
    src="https://images.unsplash.com/photo-1491528323818-fdd1faba62cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80"
  />
  <img
    alt="Gwen Stecy"
    className="inline-block h-14 w-14 rounded-full ring-2 ring-white"
    data-testid="avatar-image"
    src="https://images.unsplash.com/photo-1550525811-e5869dd03032?ixlib=rb-1.2.1&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80"
  />
  <img
    alt="Zack Snyder"
    className="inline-block h-14 w-14 rounded-full ring-2 ring-white"
    data-testid="avatar-image"
    src="https://images.unsplash.com/photo-1500648767791-00dcc994a43e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2.25&w=256&h=256&q=80"
  />
  <img
    alt="Tobey Maguire"
    className="inline-block h-14 w-14 rounded-full ring-2 ring-white"
    data-testid="avatar-image"
    src="https://images.unsplash.com/photo-1472099645785-5658abf4ff4e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80"
  />
  <div
    className="relative inline-flex h-14 w-14 items-center justify-center overflow-hidden rounded-full bg-gray-200 ring-2 ring-white dark:bg-gray-200"
  >
    <span
      className="text-xl font-medium text-gray-600 dark:text-gray-600"
      data-testid="remaining-avatar"
    >
      +3
    </span>
  </div>
</div>
`;
